src/providers/recentmessages/RecentMessagesProvider.hpp

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "common/Channel.hpp"
+#include "common/Outcome.hpp"
+#include "util/RatelimitBucket.hpp"
+
+namespace chatterino {
+
+class RecentMessagesProvider
+{
+public:
+    static RecentMessagesProvider &instance();
+
+    RecentMessagesProvider(const RecentMessagesProvider &original) = delete;
+    RecentMessagesProvider(RecentMessagesProvider &&other) = delete;
+    RecentMessagesProvider &operator=(const RecentMessagesProvider &original) =
+        delete;
+    RecentMessagesProvider &operator=(RecentMessagesProvider &&other) = delete;
+
+    void queueLoadFor(std::weak_ptr<Channel> channel);
+
+private:
+    RecentMessagesProvider();
+
+    // TODO(leon): Figure out reasonable values for these.
+    // TODO(leon): Potentially make these configurable for custom message providers
+    static constexpr int REQUEST_BUDGET{10};
+    static constexpr int REQUEST_COOLDOWN{1000};
```

Comment:
If we go the configurable route, which I don't see why not, these should be environment variables. A value of zero could disable rate limiting.

Reply:
I agree, I intend to follow the established format of the existing `CHATTERINO2_RECENT_MESSAGES_URL` here